src/app.provider.ts

```diff
@@ -0,0 +1,98 @@
+import { INestApplication, VersioningType } from '@nestjs/common';
+import { DataSourceErrorFilter } from './routes/common/filters/data-source-error.filter';
+import { DocumentBuilder, SwaggerModule } from '@nestjs/swagger';
+import { NestFactory } from '@nestjs/core';
+import { TestingModule } from '@nestjs/testing/testing-module';
+
+function configureVersioning(app: INestApplication) {
+  app.enableVersioning({
+    type: VersioningType.URI,
+  });
+}
+
+function configureShutdownHooks(app: INestApplication) {
+  app.enableShutdownHooks();
+}
+
+function configureFilters(app: INestApplication) {
+  app.useGlobalFilters(new DataSourceErrorFilter());
+}
+
+function configureSwagger(app: INestApplication) {
+  const config = new DocumentBuilder()
+    .setTitle('Safe Client Gateway')
+    .setVersion(process.env.npm_package_version ?? '')
+    .build();
+
+  const document = SwaggerModule.createDocument(app, config);
+  SwaggerModule.setup('', app, document);
+}
+
+/**
+ * The main goal of {@link AppProvider} is to provide
+ * a {@link INestApplication}.
+ *
+ * Extensions of this class should return the application in
+ * {@link getApp}.
+ *
+ * By default the {@link AppProvider} has a configuration collection ({@link setup})
+ * that can be changed by each extension of the class
+ */
+export abstract class AppProvider {
+  protected readonly setup: Array<(app: INestApplication) => void> = [
+    configureVersioning,
+    configureShutdownHooks,
+    configureFilters,
+    configureSwagger,
+  ];
+
+  public async provide(module: any): Promise<INestApplication> {
+    const app = await this.getApp(module);
+    this.setup.forEach((f) => f(app));
+    return app;
+  }
+
+  protected abstract getApp(module: any): Promise<INestApplication>;
+}
+
+/**
+ * The default {@link AppProvider}
+ *
+ * This provider should be used to retrieve the actual implementation of the
+ * service
+ */
+export class DefaultAppProvider extends AppProvider {
+  protected getApp(module: any): Promise<INestApplication> {
+    return NestFactory.create(module);
+  }
+}
+
+/**
+ * A test {@link AppProvider}
+ *
+ * This provider provides an application given a {@link TestingModule}
+ *
+ * If the module provided is not a {@link TestingModule}, an error is thrown
+ */
+export class TestAppProvider extends AppProvider {
+  constructor() {
+    super();
+    if (process.env.NODE_ENV !== 'test') {
+      throw Error('TestAppProvider used outside of a testing environment');
+    }
+
+    // Disables shutdown hooks for tests (they are not required)
+    // Enabling this in the tests might result in a MaxListenersExceededWarning
+    // as the number of listeners that this adds exceed the default
+    const index = this.setup.indexOf(configureShutdownHooks);
+    if (index > -1) this.setup.splice(index, 1);
```

Comment:
What about defining the array of setup functions explicitly for each class extending `AppProvider`? In my opinion, is a bit clearer than modifying the parent class array, but I leave this to your preference 🙂

I was thinking in something like:
```javascript
export abstract class AppProvider {
  protected setup: Array<(app: INestApplication) => void>;
  ...
```
And then: 
```javascript
export class DefaultAppProvider extends AppProvider {
  constructor() {
    super();
    this.setup = [configureVersioning, configureShutdownHooks, configureFilters, configureSwagger];
    ...
```
```javascript
export class TestAppProvider extends AppProvider {
  constructor() {
    super();
    this.setup = [configureVersioning, configureFilters, configureSwagger];
    ...
```

Reply:
> What about defining the array of setup functions explicitly for each class extending `AppProvider`? In my opinion, is a bit clearer than modifying the parent class array, but I leave this to your preference 🙂
> 
> I was thinking in something like:
> 
> ```js
> export abstract class AppProvider {
>   protected setup: Array<(app: INestApplication) => void>;
>   ...
> ```
> 
> And then:
> 
> ```js
> export class DefaultAppProvider extends AppProvider {
>   constructor() {
>     super();
>     this.setup = [configureVersioning, configureShutdownHooks, configureFilters, configureSwagger];
>     ...
> ```
> 
> ```js
> export class TestAppProvider extends AppProvider {
>   constructor() {
>     super();
>     this.setup = [configureVersioning, configureFilters, configureSwagger];
>     ...
> ```

My concern with this approach is that you stop sharing the same configuration. Ideally the `TestAppProvider` should be exactly the same as the `DefaultAppProvider`. 

If we add more configurations I see it as adding to the abstract `AppProvider` and that's it (unless it breaks the tests). Also ideally there wouldn't be a `TestAppProvider` and we would the _normal_ `DefaultAppProvider`.

Unfortunately how you create a test application is different than how you create a normal application:

```typescript
testApp = module.createNestApplication()
// vs
normalApp = NestFactory.create(module)
```

I think we can consider your approach in the future. I'm adding a new logging configuration which needs to be on both the testing and the default providers. We can see how it develops over time.